### DISCUSSION TITLE
ASL Core Board

### DISCUSSION THREAD

**Discussion by su-aac** [UNC-Project-Open-AAC/OS-DPI#243] [Ideas]
I created the 36-location core board for ASL using MP4 format. For signs that are the same, such as "he," "she," and "it," I used the same image. 

We must be cautious when using this board to ensure we don't prioritize spoken language over sign language. For example, ASL uses directional cues to convey meanings—like the phrase "help me" is expressed by moving the sign for "help" towards oneself, rather than signing the two separate words "help" and "me."
<img width="960" alt="image" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/26839abf-4f5a-43e9-8a82-2ee154608fef">
[grid_ex_1_ASL.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14425049/grid_ex_1_ASL.zip)

**Comment by aetkie** [comment]
This is a neat idea @hsiaotingsu. It's a little hard to see the videos for me only because of the small size. Do you think it would be just as effective in a 5x5 or 4x4 grid? I know 36 is a standard but visually, it's a little tough. But could just be my eyes

**Reply by gbishop** [reply]
You could make them Zoom on hover with a bit of CSS in a Customize component.

**Reply by gbishop** [reply]
Or even in the Cue (I think).

**Comment by sofia-bg** [comment]
Here is a version with zoom using CSS in the cues. Currently zoom to 2x but can change bigger or smaller. Works with mouse not sure what this looks like with touch. 

[grid_ex_1_ASL_zoom2x.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14451080/grid_ex_1_ASL_zoom2x.zip)

**Reply by gbishop** [reply]
Looks fine with touch. It ought to work with scanning also.

**Reply by sofia-bg** [reply]
@gary presumably it would still zoom with touch? but only while touching and for length of dwell time? Was thinking if you want to touch and then have the video show for some period of time after touching would need to do a modal? Or is there a way in the methods to have a "zoom" on touch and then select on a second touch? so that you could lift finger and see the video to make sure it was the one you wanted to select? If you do not select a second time would minimize after certain period of time.

**Reply by sofia-bg** [reply]
Never mind I just need to see how it behaves in touch to see how to manipulate since it is currently set to click with mouse I don't have a good sense of what that translates to with touch.

**Reply by gbishop** [reply]
My Chromebook has a touch screen. It animates while I hold over it. I had to change the activate event to Pointer Up (which we should be using for everything anyway) to get it to activate using touch. I can slide around the screen with my finger viewing the animations then release to trigger the one I want. 

It would be better (maybe?) if you could move your finger so you could see the animation... We could probably do something with the animation outside the grid. I'd have to think about how to do that. Or we could do something with a timer...

**Reply by sofia-bg** [reply]
Yeah I am just realizing I have really only been using either the mouse of the keyboard to access and haven't tried much with touch. @hsiaotingsu we can play around with it on the surface pro in the lab if you want sometime

**Comment by su-aac** [comment]
https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14852992/grid_ex_1_ASL_zoom2x.zip

**Comment by gbishop** [comment]
Here is a version that displays **multiple words as text and animated videos**. Click the text display to make it speak and clear the display. It's clunky but it appears to work. It even scrolls the animated display if you enter more than 6 words.

[grid_ex_1_ASL_zoom2x_animated_display.zip](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15249182/grid_ex_1_ASL_zoom2x_animated_display.zip)

**Reply by su-aac** [reply]
WOW! This is amazing, @gbishop !!! 
I wonder if it's possible to make it speak and at the same time play animated videos in sequence.

**Reply by gbishop** [reply]
I have played with highlighting words as they are spoken. That is in a branch in my repo. The browser provides an event that fires on each word. Unfortunately it only tells you the character offset into the string being spoken. We'd somehow have to map from the offset to the word...

**Comment by sofia-bg** [comment]
Oh cool Gary! And poking around it actually looks like less complicated actions than for the notebook. Can you explain why/how the media comes along with the label? In the actions it looks like you are saying that if you activate a button in the “TalkGrid” and the first word cell is blank then add the label to the display and the label to that first blank cell… and so on depending on which cells are blank.
 
But why does asking it for the label to the cell also pull in the video clip?
 
 
ALSO – because any customize you add to the word gird buttons doesn’t appear until a word is selected, I was able to set the background color to a light yellow and then those buttons end up being “highlighted” in a static way once filled. Thought it could be an interesting way to help distinguish between the cells that are for selecting and the ones that represent what you selected.

**Reply by gbishop** [reply]
The animated "display" is a row Stack of 1x1 Grids each filtering on a different $word1-6. A real hack! So, I assign the #label to the appropriate $word1-6 and the corresponding grid shows that row from the content.

I really like highlighting the words as they appear.

**Comment by su-aac** [comment]
I was playing with the 2-switch scanning method with Gary's design.

I found that when using the 2-switch scanning settings with yellow overlay highlighting, all videos in the top row of animated ASL would play simultaneously. And I can also select each animated ASL and make it play individually.
 
However, when using the 2-switch scanning settings with the Zoom function activated, the entire row of core words is enlarged, which covers the individual words. Additionally, it becomes challenging to tell when the selection has moved to the top row of the animated ASL video or the second row of the speech window because these areas do not zoom in accordingly.

Screen recordings
Scanning with highlight:
https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/4384e97b-5e0c-4c4c-9221-88582ca5e794 
Scanning with Zoom:
https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/3774e3a5-8035-4b80-81e0-1ece304efdeb

Can the highlight function and the Zoom function be activated for different sections when using 2-switch scanning?
For example, use the highlight function when selecting the top row of the animated ASL video, the second row of the speech window, and each row of the core words. Then, use the Zoom function to select each core word after selecting the row using the highlight function.

**Comment by gbishop** [comment]
You can tweak the highlights per component or button by adding another
attribute selector to the cue. I'm not looking at cues right now but their
syntax is something like button[cue="key"] { stuff here }

If you add another square bracket enclosed condition you can restrict it.
Maybe [#ComponentName="display"]

Bump it against the other one.

By adjusting the pattern you can control when or if buttons are visited.

Gary

**Reply by gbishop** [reply]
In the pattern tab you can set different cues for groups. So you can have it zoom on the buttons without zooming on the rows.

**Comment by su-aac** [comment]
This is awesome, Gary!

This is the screen recording of the new setting- have it highlighted on the rows and zoomed on the buttons.
https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/7cdd65bd-7b67-4881-a454-ff894e0ce7d8

However, why did the highlight scanning skip the speech window? 
Link to the board:
https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15377521/grid_ASL_scanning_zoom_highlight.zip  

[grid_ASL_scanning_zoom_highlight.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15377521/grid_ASL_scanning_zoom_highlight.zip)

**Comment by gbishop** [comment]
This is a bug. It is crashing the browser on my Chromebook. I'll look into
it.

**Reply by gbishop** [reply]
The problem may be that it is running out of memory. Activating all those videos at once may be too much.

Activating all the videos when their group is highlighted is an artifact of the hacky way I handle groups. It is not obvious to me that 36 simultaneously running videos is useful..

I'm thinking of modifying the code so that videos are played only on single buttons and not groups. What do you think?

**Reply by su-aac** [reply]
I agree with you, Gary! Running 36 videos at the same time is useless.

**Reply by gbishop** [reply]
I'm making progress. This is complicated but I think I've got it.

**Comment by gbishop** [comment]
I have modified our application and pushed it to my repo. Here is a new version of your design but, for now, it won't work on the production version. 

I used different colors for each group in the Pattern to make it clear what is happening. I'm not recommending that. Group Cues now apply to the content of the Group and not the Group itself. This is a breaking change. We've got to see if it breaks other existing designs before I push it to production.

I tweaked the Cue quite a bit! The zoomed buttons are shifted right on the left side, left on the right side and up a bit on the bottom. They are also clipped so they don't overlap the other buttons too much. I'll be happy to explain any of that. 

[grid_ASL_scanning_zoom_highlight.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15416901/grid_ASL_scanning_zoom_highlight.zip)

You can try it on my version using this [link](https://gbishop.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15416901/grid_ASL_scanning_zoom_highlight.zip)

Let me know how this works for you.

**Comment by su-aac** [comment]
Hi Gary,

The new version is neat! Thank you for the update and the detailed explanation of the changes. I appreciate your efforts in improving the application. 

From what I understand, the red overlay indicates the section (either the grid or the speaking window), the green overlay indicates a scanner to select the row, and the zoom indicates a selector. How did you tell the system to make the green overlay move by row instead of column?
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/41b56830-29b5-40b3-8945-871d9c3e92c2)

I was wondering if the scanning feature could also highlight the top row of the animated video. This would allow users who are not yet able to read words to see their ASL selections.

**Comment by gbishop** [comment]
It moves by row because I Group By #row in that Selector. If I had grouped by #column it would have moved column wise.

The red overlay gets applied to things at the top level. In this case they happen to be the TalkGrid and the display. The green overlay gets applied the content of the TalkGrid (the filter assures that), then those are Grouped By #row with Zoom applied to the members of each row. 

Yes, you can scan the top row. Just add it to the top level group. You see that Filter with #ComponentName = 'speak', just add another one to grab those little grids I used to create the Top row. I think Filter #ComponentName = 'grid' should do it. Though "grid" is a lazy name for those grids.... I was just bashing something together. You should think of a better name for them.

**Reply by su-aac** [reply]
Thanks so much, Gary! It works!!
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/6a258cbb-1d77-4011-ab4e-295293de5c5a)

**Comment by gbishop** [comment]
I'm not advocating using Red and Green like that. Make them both Yellow (or whatever). I just wanted it to distinguish the things getting selected.

**Comment by su-aac** [comment]
Is it possible that while scanning, we can hear the sound of each word before selecting it? (And also hear the first word of the row during row scanning)

File:
[grid_ASL_scanning_zoom_yellow_highlight.zip](https://github.com/user-attachments/files/16339529/grid_ASL_scanning_zoom_yellow_highlight.zip)
Link:
https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/user-attachments/files/16339529/grid_ASL_scanning_zoom_yellow_highlight.zip

**Reply by aetkie** [reply]
I think you are looking for an example of auditory scanning. We should have one. Let me look and if anyone knows of one off the top of their head, please post a link.

**Reply by gbishop** [reply]
I thought this would be easy but I see that it is not currently possible. What is your use case? That will help me think about how it should work.

**Reply by sofia-bg** [reply]
could it be a second response along with "next"? I am not sure this would be a feature we would use with deaf users, but there are users that use this feature on other device when they scan. Often it speaks the word as it scans, but quieter and maybe even in a different voice. This allows the scanner to know what word/row they are on but without others necessarily knowing/hearing. Then a different voice at a different volume would speak the word when selected. I knew someone who used this method and then did not mount their AAC device in front of them, but behind them on the back of their wheelchair. It tends to be called auditory fishing or auditory scanning.

**Reply by gbishop** [reply]
I think it should be a Cue. Cues have visual components and now they'll have audible components. I think I see how to implement it but the storm last night ruined sleep for our dog and thus for me. My brain is barely functioning. I'll have something soon.

**Reply by gbishop** [reply]
I have updated the app to support speech and sound cues. I modified your design to show how to use them. 

[grid_ASL_scan_zoom_speech.zip](https://github.com/user-attachments/files/16400777/grid_ASL_scan_zoom_speech.zip)

Let me know if you have any questions.

**Reply by su-aac** [reply]
@gbishop This is magic! Thank you, Gary. The speech and sound cues work awesome. I like the cue that said: "Core" for the whole - "Row 1/ Row 2" for each row- and then said each word out loud. And when I selected the message window, the cue said, "Speak."

I attached a video of me using the board.
https://github.com/user-attachments/assets/4bff8026-07fe-4c0b-9887-b0d1e345c043

**Reply by gbishop** [reply]
Examine the Pattern and the Cues and you'll be able to see why it announces the various elements. You have complete control over that.

**Reply by gbishop** [reply]
I suggest you tweak the voice used for the Cues to make it sound different. I left it at the default so it would work anywhere but even with that you can adjust the volume, rate and pitch.

**Comment by su-aac** [comment]
This is really cool! You can swipe your finger to scan, and when you release your finger, it will make a selection. The auditory scanning happens while your finger is sliding.

Video demo of me using my finger to directly select symbols on the board:
https://github.com/user-attachments/assets/71b01057-e389-4089-9654-6423d84ee9a9

Screenshot of the Methods- remember to select "Pointer up" so the direct selection works.
![image](https://github.com/user-attachments/assets/05435442-8382-4a70-b636-8b8982ebf8f2)

**Comment by sofia-bg** [comment]
Very cool!

From: Sophie Su ***@***.***>
Date: Tuesday, August 13, 2024 at 12:16 PM
To: UNC-Project-Open-AAC/OS-DPI ***@***.***>
Cc: Sofia B-G ***@***.***>, Comment ***@***.***>
Subject: Re: [UNC-Project-Open-AAC/OS-DPI] ASL Core Board (Discussion #243)

This is really cool! You can swipe your finger to scan, and when you release your finger, it will make a selection. The auditory scanning happens while your finger is sliding.

Video demo of me using my finger to directly select symbols on the board:
https://github.com/user-attachments/assets/71b01057-e389-4089-9654-6423d84ee9a9

Screenshot of the Methods- remember to select "Pointer up" so the direct selection works.
image.png (view on web)<https://github.com/user-attachments/assets/05435442-8382-4a70-b636-8b8982ebf8f2>
